fs: added usage examples/doctests to File

mattgathu opened the pull request:
Motivation
--
Part of [this](https://github.com/rust-lang-nursery/wg-net/issues/54) networking working group issue.

Solution
--
Add doctests to public methods of the `tokio::fs::File` struct.

- file::File
- file::File::from_std
- file::File::create
- file::File::into_std
- file::File::metadata
- file::File::open
- file::File::poll_metadata
- file::File::poll_seek
- file::File::poll_set_len
- file::File::poll_set_permissions
- file::File::poll_sync_all
- file::File::poll_sync_data
- file::File::poll_try_clone
- file::File::seek

tobz commented on tokio-fs/src/file/mod.rs:
```diff
@@ -34,6 +34,46 @@ use std::path::Path;
+/// fn main() {
+///     let task = tokio::fs::File::open("foo.txt")
+///         .and_then(|mut file| {
+///             let mut contents = vec!();
```
I think you meant `vec![]`?

tobz commented on tokio-fs/src/file/mod.rs:
````diff
@@ -34,6 +34,46 @@ use std::path::Path;
+/// ```no_run
+/// extern crate tokio;
+/// 
+/// use tokio::prelude::*;
````
Do you actually need the prelude here?  Given that the File type is fully qualified?

tobz commented on tokio-fs/src/file/mod.rs:
```diff
@@ -73,15 +130,45 @@ impl File {
+    /// # extern crate tokio;
+    /// use tokio::prelude::*;
+    /// fn main() {
+    ///     let task = tokio::fs::File::create("foo.txt").and_then(|file| {
```
Break the `and_then` to its own line.

tobz commented on tokio-fs/src/file/mod.rs:
```diff
@@ -73,15 +130,45 @@ impl File {
+    /// # extern crate tokio;
+    /// use std::fs::File;
+    /// 
+    /// fn main() -> Result<(), Box<std::error::Error>> {
```
I think drop the `Result` (and the corresponding `Ok`) from this.  We aren't going to run it in CI, so we can be a little more terse.

mattgathu replied:
cool. I have dropped the result and now just have a single unwrap.

tobz commented on tokio-fs/src/file/mod.rs:
```diff
@@ -143,23 +310,96 @@ impl File {
+    ///         # println!("{:?}", clone);
+    ///     }).map_err(|err| eprintln!("IO error {:?}", err));
+    /// 
+    /// tokio::run(task);
```
Indent.

tobz commented on tokio-fs/src/file/mod.rs:
```diff
@@ -182,6 +422,29 @@ impl File {
+    ///     let task = tokio::fs::File::create("foo.txt")
+    ///         .and_then(|file| file.metadata())
+    ///         .map(|file_metadata| {
+    ///             let (mut file, metadata) = file_metadata;
```
I might collapse this line and pull out both `file` and `metadata` right from the closure params.

tobz commented on tokio-fs/src/file/mod.rs:
```diff
@@ -52,6 +92,23 @@ impl File {
+    ///         // do something with the file ...
+    ///         file.metadata().map(|md| println!("{:?}", md))
+    ///     }).map_err(|e| {
+    ///         //handle errors
```
Space between `//` and sentence.

tobz commented on tokio-fs/src/file/mod.rs:
```diff
@@ -73,15 +130,45 @@ impl File {
+    ///         // do something with the created file ...
+    ///         file.metadata().map(|md| println!("{:?}", md))
+    ///     }).map_err(|e| {
+    ///         //handle errors
```
Space between `//` and sentence.

mattgathu commented:
@tobz I have updated the pr:

- used specific imports from `prelude` instead of a wildcard.
- refactored IO error `map_err` and `eprintln!`
- refactored out `Result` in one of the examples

[diff](
https://github.com/tokio-rs/tokio/pull/786/commits/31b1bb0a56b387e34ac44ac441dd3f7ceac4c189)

tobz commented:
Great, I think things are consistent enough now. 👍
